specs/CustomDownload.md

```diff
@@ -0,0 +1,784 @@
+    // Selecting `OK` will save the download to the chosen path.
+    // Selecting `CANCEL` will cancel the download.
+    CHECK_FAILURE(m_webView->add_DownloadStarting(
+        Callback<ICoreWebView2DownloadStartingEventHandler>(
+            [this](
+                ICoreWebView2_3* sender,
+                ICoreWebView2DownloadStartingEventArgs* args) -> HRESULT
+            {
+                auto showDialog = [this, args]
+                {
+                    wil::com_ptr<ICoreWebView2Download> download;
+                    CHECK_FAILURE(args->get_Download(&download));
+
+                    INT64 expectedDownloadSizeInBytes = 0;
+                    CHECK_FAILURE(download->get_ExpectedDownloadSizeInBytes(
+                      &expectedDownloadSizeInBytes));
+
+                    wil::unique_cotaskmem_string uri;
+                    CHECK_FAILURE(download->get_Uri(&uri));
+
+                    wil::unique_cotaskmem_string mimeType;
+                    CHECK_FAILURE(download->get_MimeType(&mimeType));
+
+                    wil::unique_cotaskmem_string contentDisposition;
+                    CHECK_FAILURE(download->get_ContentDisposition(&contentDisposition));
+
+                    // Get the suggested path from the event args.
+                    wil::unique_cotaskmem_string resultFilePath;
+                    CHECK_FAILURE(args->get_ResultFilePath(&resultFilePath));
+
+                    std::wstring prompt =
+                        std::wstring(L"Enter result file path or select `OK` to use default path. ") +
+                        L"Select `Cancel` to cancel the download.";
+
+                    std::wstring description = std::wstring(L"URI: ") + uri.get() + L"\r\n" +
+                                                L"Mime type: " + mimeType.get() + L"\r\n" +
+                                                L"Expected Download size in bytes: " +
+                                                std::to_wstring(expectedDownloadSizeInBytes) +
+                                                L"\r\n";
+
+                    TextInputDialog dialog(
+                        m_appWindow->GetMainWindow(), L"Download Starting", prompt.c_str(),
+                        description.c_str(), resultFilePath.get());
+                    if (dialog.confirmed)
+                    {
+                        // If user selects `OK`, the download will complete normally.
+                        // Result file path will be updated if a new one was provided.
+                        CHECK_FAILURE(args->put_ResultFilePath(dialog.input.c_str()));
+                        UpdateProgress(download.get());
+                    }
+                    else
+                    {
+                        // If user selects `Cancel`, the download will be canceled.
+                        CHECK_FAILURE(args->put_Cancel(TRUE));
+                    }
+                };
+
+                wil::com_ptr<ICoreWebView2Deferral> deferral;
```

Comment:
```suggestion
                // Obtain a deferral for the event so that the CoreWebView2
                // doesn't examine the properties we set on the event args until
                // after we call the Complete method asynchronously later.
                wil::com_ptr<ICoreWebView2Deferral> deferral;
```